### DISCUSSION TITLE
Request chaining on ad hoc filters in Grafana

### DISCUSSION THREAD

**Discussion by garrettlish** [grafana/grafana#45363] [Data Sources]
I have a dashboard in which I want to use an ad hoc filter. Once I selected one label key and value (f.e. labelKey1 and labelValue1), while selecting next label key and value (f.e. labelKey2 and labelValue2), it doesn't exclude all of the labelset which labelKey1!=labelValue1.

Use loki data source as an example: can we use label API to fetch the first label key and value, and use series API to fetch the follow-up label keys and values?

**Why is this needed:**
It could significantly reduce the follow-up label keys and values possible values, make the ad hoc filters more efficient.

Any thoughts, please advise, thanks!

**Comment by v-zhuravlev** [comment]
We use variables chaining technique extensively in Grafana cloud-integrations.

I like ad-hoc filter idea. We can add to all our integrations as the last selector in variables chain. However, currently it lacks the feature to filter output based on what is already being selected in other templated variables.If this is implemented, Our default variables chain can be something like

    datasource
    filtering selector like (job="integrations/node_exporter")
    group selector, like (job=~$job)
    instance selector like (instance=~$instance)
    ad-hoc

where ad-hoc results are filtered by what was chosen in filter/job/instance.For example, if job is set to 'All' and instance is set to 'All', this will effectively narrow down labels only to ones available in Linux integration (job="integrations/node_exporter").This should be very useful when users have some custom external labels added(like cluster, zone, environment...)  and what to filter by them, not job/instance available OOTB on our dashboards.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
